### ISSUE TITLE
[TECH-DEBT] - Handle multivalue support for www urlencoded header

### ISSUE THREAD

**Issue by authorjapps** [authorjapps/zerocode#148]
This is a tech-debt ticket.
Before implementing, please check if the framework(Apache Http Client) is **already taking care** of this situation. If not, then implement it and raise a PR.
TODO-
+ Add an end point or use the existing end point in the Spring Boot demo repo to reproduce this
   + See the relevant repo/class in the below section
+ Add a simple brand new test-case with couple of Key-Values to the above repo
+ Please take care of "body" and "queryParams" while fixing the same.
   + The fix is similar for both. You might reuse the same code.

Spring boot repo:
https://github.com/authorjapps/spring-boot-integration-test

Controller class:
https://github.com/authorjapps/spring-boot-integration-test/blob/master/src/main/java/com/springboot/controller/CustomerController.java

**_Relevant comments from the old ticket:_**
Comment Link: https://github.com/authorjapps/zerocode/issues/141#issuecomment-432700435

That sounds correct. 👍 This will make everyone's life a lot easy now. 
Postman accepts both as `key-value` and combines them if both are present(shown below).

Folks, Let's go with **PR review n approval** please !

![postmna_body](https://user-images.githubusercontent.com/12598420/47440299-c507d280-d7a5-11e8-9f3b-9bc795d9ef14.png)

- Then in server app-
![postman_rq_rcvd](https://user-images.githubusercontent.com/12598420/47440323-d4871b80-d7a5-11e8-9541-13b40de73d90.png)

If **same key** is present, they server receives as multivalue(means client also sends as multi value) - see below:
- postman client
![postman_2vals](https://user-images.githubusercontent.com/12598420/47440865-f503a580-d7a6-11e8-9679-6565a6417463.png)

- spring boot server
![postman_2vals_server](https://user-images.githubusercontent.com/12598420/47440892-fe8d0d80-d7a6-11e8-844f-a81142fd60dd.png)

@arunvelusamyd , Good job mate ! Have you taken care of the above(_same key situation_) in your PR ? If not can you please take care now ? 👍 

@macrocks , @cooljavadev @nirmalchandra - Please review n approve -or- express your view points if disagree or think any other way  🙏

**Comment by pgmpofu** [comment]
@authorjapps  May I work on this?

**Comment by authorjapps** [comment]
Yes, sounds good. 👍 

Greetings and Welcome, Thanks for your interest in contributing.

+ Mailing list - Can you please join the mailing list?
   + https://groups.google.com/forum/#!forum/zerocode-automation

+ Please visit this(2min read) CONTRIBUTION guidelines page
   + https://github.com/authorjapps/zerocode/blob/master/CONTRIBUTING.md

+ Gitter link acts as QnA forum currently, until we have all members in the mailing-list above(newly created)
    + https://gitter.im/zerocode-testing/help-and-usage

Cheers

**Comment by pgmpofu** [comment]
@authorjapps  I apologize for the inconvenience but I am unable to work on this due to other commitments. Please feel free to reassign.

**Comment by authorjapps** [comment]
No worries. Thanks for letting us know. 👍

**Comment by oiakinat** [comment]
@authorjapps is this issue still relevant? May I work on it?

**Comment by authorjapps** [comment]
Hello mate, I have added more details(springboot repo, controller class) to the description now.
Yes please, you can kindly pick this.

And thanks for your participation in contributing!

And if you could please join these-

+ Mailing list - Can you please join the mailing list?
   + https://groups.google.com/forum/#!forum/zerocode-automation

+ Please visit this(2min read) CONTRIBUTION guidelines page
   + https://github.com/authorjapps/zerocode/blob/master/CONTRIBUTING.md

+ Gitter link acts as QnA forum 
    + https://gitter.im/zerocode-testing/help-and-usage

Cheers

**Comment by oiakinat** [comment]
@authorjapps Thanks a lot for this great opportunity) I've already done all these instructions and some research with fixes.
So, I've figured out that SpringBoot recognize same key from query parameters and HttpEntity as it was described  in comments from #141 
For example, ZeroCode JSON with these objects `"queryParams": {"notArray": "notOne"},"body": {"notArray": "notTwo"}` will be recieved by SpringBoot as `{notArray=[notOne, notTwo]}`
But I guess it'll be nice, if we could use arrays in `"queryParams"` and `"body"`, because now, if we'll try to send something like this `"queryParams": {"array": ["one","two"]},"body": {"array": ["three","four"]}` SpringBoot will receive `{array=[[one, two],[three,four]]}`. So I've fixed it, and now ZeroCode support multivalue not only for www-urlencoded header. If it's ok, can I raise a PR? Or if it's not ok - could you please tell where I've got it wrong?
Also, I have to say that in my fix I can't see really good way to control value duplicates, but should we really do this?

**Comment by authorjapps** [comment]
Solution sounds great. Thank you. We are reviewing this and will come back here soon.

@arunvelusamyd @santhoshTpixler @macrocks @aghorpade @officiallysameer @respondsid

**Comment by santhoshTpixler** [comment]
Mixing the parameters in the URL(AKA queryParameters) and the body parameters will introduce more confusion. It is better to have a clear separation between the URL and the body.

**Comment by oiakinat** [comment]
I agree, but in my comment using "queryParams" and "body" together, just a demonstration. Of course, they can be used separately - my idea was to give an additional opportunity to use arrays for params both in "queryParams" and in "body".

**Comment by priya1712d** [comment]
@authorjapps Hello! I am interested in working on this issue. Is anyone working on this?

**Comment by authorjapps** [comment]
@priya1712d , please go ahead and raise a PR for this when ready.  

Please have a look here for CONTRIBUTING:
https://github.com/authorjapps/zerocode/blob/master/CONTRIBUTING.md

**Comment by JoaoFelipe76** [comment]
Can I work on this, @authorjapps?
